### PR TITLE
Set -g, record backtraces by default

### LINKED ISSUE #11278: Add an hint suggesting record backtrace when a program fail because of an uncaght exception
I asked [on discuss OCaml](https://discuss.ocaml.org/t/why-isnt-ocaml-recording-bactrace-by-default/9915/5) why the backtrace wasn't recorded by default. I was told it was because of a performance issue.

However, this is the kind of features that make the life of beginners way easier. Maybe that the compiler could give them a hint like 
> You can try to use `Printexc.record_backtrace true` or `OCAMLRUNPARAM="b"` to get the trace of this error

when an uncaught exception is risen

### PR DESCRIPTION
**Pull request by nojb**
This PR makes the `-g` flag enabled by default, as well as the recording of backtraces (normally controlled by `Printexc.record_backtrace` and/or `OCAMLRUNPARAM`).

A flag `-no-g` is added in order to disable the generation of debugging information.

Fixes #6728
Fixes #11278

### CODE REVIEW

**Comment by gasche** [comment]
Making an option the default and adding the anti-option in the same release is a compatibility antipattern, because people cannot enable the option without breaking the build for older language releases. On the other hand, our expectation is that essentially no one cares in practice about `-no-g`, because `-g` itself should have no essentially performance impact -- adding debug information without backtraces only changes code generation slightly, and makes binaries bigger. (This might be an issue in some very specialized settings, and those maybe will deal fine with having to use a fixed OCaml version.)

The real performance impact comes from enabling backtrace, in combination with usage of exception for control flow, unless `raise_notrace` is specifically used for those. For this "fix" is to reset `OCAMLRUNPARAM="b=0"`, and this is backward-compatible with older releases.

In short: the details are a bit lousy, but I think the overall change is fine, and I'm definitely in favor of this usability improvement.

**Comment by jberdine** [comment]
If I remember my experience correctly, "once upon a time" adding `-g` could affect what code ocamlopt generated, beyond the inclusion of debug info. The last time I check that carefully was many years ago, so perhaps that is no longer the case. I made a quick scan of the compiler just with grep, etc. and did not find anything, but it would be good to have people familiar with the middle ends and back ends comment affirmatively that it would be considered an unknown bug if adding `-g` to ocamlopt changed the code that was generated.

Separately, I must say that I find the option to give a hint that backtraces could be produced when an unhandled exception is reported overall more discoverable and user-friendly. Yes, for people who are just starting it would be one additional cycle to run their program and get a hint to set OCAMLRUNPARAM and rerun, but that is likely to happen with small short-running programs that are likely to undergo many iterations anyhow. On the other hand, if performance is silently degraded for all code by default, unless a programmer already knows enough to disable the backtraces, then it is very hard to discover this. I think that overall people who ask about such things on discuss get good answers, which is great, but how many people don't know to ask, or have some hesitation to do so? Maybe it would be different if exceptions were a new or almost unused feature, but they are not: there is a huge amount of code that uses `raise` rather than the much more recent `raise_notrace` when what is really needed is `goto` or a one-shot continuation.

**Comment by jberdine** [comment]
Irrespective of what gets decided about defaults, it would be very good to have `-no-g`. Every default setting should have a way to explicitly set it. If a setting having the default value matters in some context, it should be set explicitly in that context.

**Comment by lthls** [comment]
> it would be good to have people familiar with the middle ends and back ends comment affirmatively that it would be considered an unknown bug if adding `-g` to ocamlopt changed the code that was generated.

There aren't many things in `ocamlopt` that depend on the `-g` flag.
As far as I can tell, in addition to the obvious generation of debug symbols, they are:
- `raise` is translated into `raise_notrace` if debug is off
- the code for array bound errors is duplicated (once for each array non-unsafe access) if debug is on, but shared at the function level if it is off
- builtin methods are not used if debug is on (my opinion is that they shouldn't be used if debug is off either, but that's another discussion)

**Comment by xavierleroy** [comment]
Enabling backtrace recording by default, despite the performance overhead, requires a serious discussion and collective approval, beyond what is normally done for a PR.  There's no way this PR can just be accepted and merged.

Enabling `-g` by default is less controversial.  Nonetheless, `-g` turns off some of the "let" simplifications when compiling to bytecode, so some measurements are in order too.

**Comment by xavierleroy** [comment]
Also, in bytecode, `-g` increases the size of object files and executables tremendously, because it records not just call site information but full typing information about variables and expressions.  I don't think this can be made the default.  An intermediate debugging level should be introduced that just records call site information.

Somewhat relatedly, `-no-g` is `-g0` in GCC and Clang.

So we could have
* `-g0` : no debugging info
* `-g1` (as a possible default): just call site information (for ocamlc and ocamlopt)
* `-g2`: full debugging info (for ocamlc only).

`-g` would be synonymous for `-g2` in ocamlc and `-g1` in ocamlopt.

**Comment by nojb** [comment]
> Also, in bytecode, `-g` increases the size of object files and executables tremendously, because it records not just call site information but full typing information about variables and expressions. I don't think this can be made the default. An intermediate debugging level should be introduced that just records call site information.

Thanks for the idea. I did some preliminary experiments (assuming I didn't miss anything):

- With `-g1`, the size of `ocamlc` is ~12MB (vs ~20MB with full debug info and ~3MB for no debug info)

- To get a rough idea of the **performance** penalty incurred when building with `-g`, I timed the compilation of `typecore.cmx` (10 times in a row) with `ocamlopt` in both bytecode and native-code, built with and without `-g`. Roughly speaking, in bytecode compiling with `-g` had a performance cost of ~5%. There was no difference in native code.

  `ocamlopt` (bytecode) with `-g`
  ```
  real    0m31.387s
  user    0m31.019s
  sys     0m0.367s
  ```
  `ocamlopt` (bytecode) without `-g`
  ```
  real    0m29.431s
  user    0m28.957s
  sys     0m0.472s
  ```
  `ocamlopt` (native code) with `-g`
  ```
  real    0m7.699s
  user    0m7.301s
  sys     0m0.397s
  ```
  `ocamlopt` (native code) without `-g`
  ```
  real    0m7.618s
  user    0m7.228s
  sys     0m0.387s
  ```

**Comment by shakthimaan** [comment]
> A flag -no-g is added in order to disable the generation of debugging information.

Can you please let me know as to where to pass the `-no-g` flag to disable the generation of the debugging information? 

I tried to pass it with `configure`, but, I get a `configure: error: unrecognized option: -no-g`. I looked at `configure` at https://github.com/nojb/ocaml/blob/g_default/configure, and I could not see it. Thanks!

**Comment by lthls** [comment]
> Can you please let me know as to where to pass the `-no-g` flag to disable the generation of the debugging information?

It's a compiler flag, not a configure flag. If you want to disable generation of debugging information globally, you can set `OCAMLPARAM` to `_,g=0` (or `g=0,_` if you only want to disable it unless explicitly added with `-g`).

**Comment by shindere** [comment]
Shakthi Kannan (2022/06/01 04:09 -0700):
> > A flag -no-g is added in order to disable the generation of debugging information.
>
> Can you please let me know as to where to pass the `-no-g` flag to disable the generation of the debugging information?
>
> I tried to pass it with `configure`, but, I get a `configure: error:
> unrecognized option: -no-g`. I looked at `configure` at
> https://github.com/nojb/ocaml/blob/g_default/configure, and I could
> not see it. Thanks!

At the moment there is, sadly, no way to pass it ot the whole build
system andyou have to modify the makefiles themselves.

There is a work in progress to make such things simpler.

The aim is, once this work is done, that you should be able to write

```
./configure OCAML_COMMON_CFLAGS=-g0
make
```

to achieve what you want, assuming that the `-g0` option has been added
to the compilers.

**Comment by nojb** [comment]
This is too big a change to make in one step; we should proceed in more incremental steps. Closing.
